### ISSUE TITLE
Decouple segment/PSD step and songdkl calculate / numsyls

### ISSUE THREAD

**Issue by NickleDave** [NickleDave/songdkl#47]
Right now, every time we run `songdkl calculate` or `songdkl numsyls`, we prepare an entire dataset that includes segmenting audio into syllables and then computing power spectral densities for each of those syllables.  

This takes a lot of time, impairing the ability to rapidly test and iterate.  
It also impairs our ability to directly test different segmenting algorithms as in #37, because we don't capture the segmenting + resulting PSDs in a way that lets us easily do an analysis.
By the same token, we can't easily investigate how segmenting is affecting downstream computations.

Given this, I propose we decouple the segment / PSD step from the functions that calculate Song_D_{KL} and estimate the number of syllables.  
@dgmets @sthaar @yangzheng-121 please let me know what you think of this idea.  
This might at first seem like adding a new feature but I think it will actually make it easier in the long run to make what's described in the paper more accessible.

At a high level, the way to do this would be to:  
* add a separate `songdkl prep` command to the command-line interface, that outputs the prepared dataset (below)
* then have the `songdkl.calculate` and `songdkl.numsyls` functions accept paths to whatever the output of `songdkl prep` is

I think the simplest way to do this for now would be to just use `zarr` to persist all the arrays, including both the segments we get back from `findobjects` and the PSDs.
https://zarr.readthedocs.io/en/stable/tutorial.html#persistent-arrays
https://zarr.readthedocs.io/en/stable/tutorial.html#storage-alternatives
Long term, it would be nice to have something more like a dataset format, but for now I just want to decouple these two steps in a way that makes it easier to dig into what's going on.

So basically, `songdkl.calculate` and `songdkl.numsyls` would accept paths to persisted `zarr` arrays, i.e., the default `DirectoryStore` one gets from `zarr.open`

Just to drive home the point of what this would enable (and so we don't forget about it)
@sthaar and @yangzheng-121 have provided some code that give an idea of the kinds of analyses + visualizations we would want to do if we could decouple the segmentation step, here:  
https://github.com/NickleDave/songdkl/issues/30#issuecomment-1238619147  
https://github.com/sthaar/audio-analyses/blob/main/songdkl_csv_onset-offset_share.py  
https://github.com/sthaar/audio-analyses/blob/main/get_spectrogram_with_lines_clean.ipynb

Note this is also related to #26, since it would be part of breaking up steps of entire pipeline.

**Comment by sthaar** [comment]
Makes sense to me! Combining steps afterwards is easier that finding out how to split them up. Proper segmentation is necessity for the rest of the analyses. Separation of the segmentation indeed makes troubleshooting/ fine tuning more efficient. And I guess it makes the computation more efficient too, as I understand syllable boundaries now computed twice (correct me if i'm wrong of course). I have no experience with zarr, but quick scan looks good to me. 

Only one thing to maybe take into account is to give an estimate of how much storage space would be needed. 

Maybe it's also an idea to implement an option to test on a small subset (e.g. 10% of the data), to speed up the testing and visualize testing? (Of course you can always do this manually)

**Comment by NickleDave** [comment]
> Makes sense to me!

:+1: 

> Only one thing to maybe take into account is to give an estimate of how much storage space would be needed.

Good point. Since it fits in memory I don't think it will be gigantic though.

> Maybe it's also an idea to implement an option to test on a small subset

Also a very good point -- in fact I should have thought of this already. Need to tinker to find out how much I can shrink the dataset size but will do this as I write scripts to set up for tests by downloading the PlosCompBio dataset locally and then making a "quick tests" dataset from those

**Comment by dgmets** [comment]
This seems entirely sensible!!  All of the foks using EVsonganaly use a
standardized file to contain offsets, onsets, etc...
